core/src/main/java/org/elasticsearch/gateway/DanglingIndicesState.java

```diff
@@ -104,36 +115,28 @@ void findNewAndAddDanglingIndices(MetaData metaData) {
      * that have state on disk, but are not part of the provided meta data, or not detected
      * as dangled already.
      */
-    Map<String, IndexMetaData> findNewDanglingIndices(MetaData metaData) {
-        final Set<String> indices;
+    Map<Index, IndexMetaData> findNewDanglingIndices(MetaData metaData) {
+        final Set<String> excludeIndexPathIds = new HashSet<>(metaData.indices().size() + danglingIndices.size());
+        for (ObjectCursor<IndexMetaData> cursor : metaData.indices().values()) {
+            excludeIndexPathIds.add(cursor.value.getIndex().getUUID());
+        }
+        excludeIndexPathIds.addAll(danglingIndices.keySet().stream().map(Index::getUUID).collect(Collectors.toList()));
         try {
-            indices = nodeEnv.findAllIndices();
-        } catch (Throwable e) {
+            final List<IndexMetaData> indexMetaDataList = metaStateService.loadIndicesStates(excludeIndexPathIds::contains);
+            Map<Index, IndexMetaData> newIndices = new HashMap<>(indexMetaDataList.size());
+            indexMetaDataList.stream()
+                .filter(indexMetaData ->
+                    metaData.hasIndex(indexMetaData.getIndex().getName()) == false && danglingIndices.containsKey(indexMetaData.getIndex()) == false)
```

Comment:
we should log a warn message here. I think it will all be cleaner to read without the filtering fancy pancing :)